File: client/session.go
package main

import (
	"bufio"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"net"
	"strings"
)

func read_server(server net.Conn, state *clientState) {
	scanner := bufio.NewScanner(server)
	configureScanner(scanner)
	for scanner.Scan() {
		var packet Packet
		if err := json.Unmarshal(scanner.Bytes(), &packet); err != nil {
			continue
		}
		if err := validatePacket(packet); err != nil {
			fmt.Println("invalid server packet")
			continue
		}
		processServerPacket(packet, state, server)
	}
	if err := scanner.Err(); err != nil {
		fmt.Println("server connection closed:", err)
	}
}

func processServerPacket(packet Packet, state *clientState, server net.Conn) {
	switch packet.Type {
	case "users":
		state.mu.Lock()
		active := make(map[string]bool, len(packet.Users))
		for _, peer := range packet.Users {
			active[peer.Username] = true
			if peer.Identity != "" {
				if old, exists := state.identities[peer.Username]; exists && old != "" && old != peer.Identity {
					state.keyChanged[peer.Username] = true
				}
				state.identities[peer.Username] = peer.Identity
			}
			state.peers[peer.Username] = peer
		}
		for name := range state.peers {
			if !active[name] {
				delete(state.peers, name)
			}
		}
		select {
		case state.peerUpdates <- struct{}{}:
		default:
		}
		state.mu.Unlock()
		usernames := make([]string, 0, len(packet.Users))
		for _, peer := range packet.Users {
			usernames = append(usernames, peer.Username)
		}
		fmt.Println("Users:", strings.Join(usernames, ", "))
	case "waiting":
		fmt.Printf("Waiting for %s...\n", packet.To)
	case "message":
		envelope, err := decryptChatEnvelope(state, packet.Payload)
		if err != nil {
			fmt.Println("decrypt error:", err)
			return
		}
		if envelope.Cover {
			return
		}
		fmt.Printf("\n[%s] %s\n> ", envelope.Sender, envelope.Body)
	case "session_offer":
		privateKey, err := ensurePrivateKey(state)
		if err != nil {
			fmt.Println("session offer keygen error:", err)
			return
		}
		shared, err := deriveSessionKey(privateKey, packet.PublicKey)
		if err != nil {
			fmt.Println("session offer key error:", err)
			return
		}
		offerBytes, err := decryptBytesAAD(shared, packet.Payload, []byte("yori/session/v1/offer"))
		if err != nil {
			fmt.Println("session offer decrypt error:", err)
			return
		}
		var offer sessionEnvelope
		if err := json.Unmarshal(offerBytes, &offer); err != nil || offer.Sender == "" || offer.PublicKey == "" || offer.SessionID == "" {
			fmt.Println("invalid session offer")
			return
		}
		state.mu.RLock()
		peerIdentity, known := state.identities[offer.Sender]
		changed := state.keyChanged[offer.Sender]
		state.mu.RUnlock()
		if changed {
			fmt.Printf("\n[warn] session offer from %s blocked: identity key changed\n> ", offer.Sender)
			
			state.mu.RLock()
			username := state.username
			state.mu.RUnlock()
			
			rejectMsg := Packet{Type: "session_reject", From: username, To: offer.Sender, Payload: "identity_key_changed"}
			rejectBytes, _ := json.Marshal(rejectMsg)
			paddedBytes := pad256(rejectBytes)
			ciphertext, ephPub, err := encryptPreSession(packet.PublicKey, paddedBytes)
			if err == nil {
				_ = directSendToUser(state, offer.Sender, Packet{Type: "deliver", To: offer.Sender, Payload: ciphertext, PublicKey: ephPub})
			}
			return
		}
		if known && peerIdentity != "" && !verifySessionSignature(peerIdentity, offer.Sender, offer.SessionID, offer.PublicKey, offer.Signature) {
			fmt.Printf("\n[warn] session offer from %s blocked: signature invalid\n> ", offer.Sender)
			return
		}
		sendKey, receiveKey, err := deriveDirectionalKeys(shared, offer.SessionID, false)
		if err != nil {
			fmt.Println("session key derivation error:", err)
			return
		}
		state.mu.Lock()
		state.keys[offer.Sender] = receiveKey
		state.outgoing[offer.Sender] = sendKey
		state.shared[offer.Sender] = shared
		state.sessionIDs[offer.Sender] = offer.SessionID
		state.mu.Unlock()
		state.mu.RLock()
		username := state.username
		state.mu.RUnlock()
		replyPublicKey := base64.RawStdEncoding.EncodeToString(privateKey.PublicKey().Bytes())
		identity, err := ensureIdentityKey(state)
		if err != nil {
			fmt.Println("identity key error:", err)
			return
		}
		reply, err := json.Marshal(sessionEnvelope{Sender: username, PublicKey: replyPublicKey, SessionID: offer.SessionID, Signature: sessionSignature(identity, username, offer.SessionID, replyPublicKey)})
		if err != nil {
			fmt.Println("session reply encoding error:", err)
			return
		}
		encryptedBobKey, err := encryptBytesAAD(shared, reply, []byte("yori/session/v1/reply"))
		if err != nil {
			fmt.Println("session reply encryption error:", err)
			return
		}
		innerReply := Packet{Type: "session_reply", To: offer.Sender, Payload: encryptedBobKey}
		innerBytes, err := json.Marshal(innerReply)
		if err != nil {
			fmt.Println("inner reply marshal error:", err)
			return
		}
		paddedBytes := pad256(innerBytes)
		ciphertext, ephPub, err := encryptPreSession(packet.PublicKey, paddedBytes)
		if err != nil {
			fmt.Println("pre-session encryption error:", err)
			return
		}
		if err := directSendToUser(state, offer.Sender, Packet{Type: "deliver", To: offer.Sender, Payload: ciphertext, PublicKey: ephPub}); err != nil {
			fmt.Println("session reply send error:", err)
			return
		}
		fmt.Printf("\nsession established with %s — type /chat %s\n> ", offer.Sender, offer.Sender)
	case "session_reply":
		sender, sessionID, publicKey, signature, err := decryptSessionReply(state, packet.Payload)
		if err != nil {
			fmt.Println("session reply decrypt error:", err)
			return
		}
		state.mu.RLock()
		identity, known := state.identities[sender]
		changed := state.keyChanged[sender]
		state.mu.RUnlock()
		if changed {
			fmt.Printf("\n[warn] session reply from %s blocked: identity key changed\n> ", sender)
			return
		}
		if known && identity != "" && !verifySessionSignature(identity, sender, sessionID, publicKey, signature) {
			fmt.Printf("\n[warn] session reply from %s blocked: signature invalid\n> ", sender)
			return
		}
		state.mu.Lock()
		if shared, ok := state.shared[sender]; ok {
			if sendKey, receiveKey, keyErr := deriveDirectionalKeys(shared, sessionID, true); keyErr == nil {
				state.outgoing[sender] = sendKey
				state.keys[sender] = receiveKey
			}
		}
		state.mu.Unlock()
		fmt.Printf("\nsession established with %s — type /chat %s\n> ", sender, sender)
	case "session_reject":
		if packet.Payload == "identity_key_changed" {
			state.mu.RLock()
			username := state.username
			state.mu.RUnlock()
			fmt.Printf("\n[warn] session rejected by %s: identity key changed on their end — ask them to run /trustkey %s\n> ", packet.From, username)
		}
	case "relay":
		if err := send_packet(server, packet); err != nil {
			fmt.Println("relay send error:", err)
			return
		}
	case "error":
		fmt.Println("Server:", packet.Payload)
	case "deliver":
		state.mu.RLock()
		preSessionKey := state.preSessionKey
		sessionPrivKey := state.privateKey
		state.mu.RUnlock()

		var plaintext []byte
		var err error
		if preSessionKey != nil {
			plaintext, err = decryptPreSessionWithKey(preSessionKey, packet.PublicKey, packet.Payload)
		}
		if (err != nil || preSessionKey == nil) && sessionPrivKey != nil {
			plaintext, err = decryptPreSessionWithKey(sessionPrivKey, packet.PublicKey, packet.Payload)
		}
		if err != nil {
			fmt.Println("pre-session decrypt error:", err)
			return
		}
		
		unpadded, err := unpad256(plaintext)
		if err != nil {
			return
		}

		var inner Packet
		if err := json.Unmarshal(unpadded, &inner); err != nil {
			fmt.Println("pre-session unmarshal error:", err)
			return
		}
		if inner.Type == "cover" || inner.IsCover {
			return
		}
		processServerPacket(inner, state, server)
	}
}

func decryptChatEnvelope(state *clientState, ciphertext string) (chatEnvelope, error) {
	state.mu.RLock()
	keys := make(map[string][]byte, len(state.keys))
	for sender, key := range state.keys {
		keys[sender] = append([]byte(nil), key...)
	}
	sessionIDs := make(map[string]string, len(state.sessionIDs))
	for sender, sessionID := range state.sessionIDs {
		sessionIDs[sender] = sessionID
	}
	state.mu.RUnlock()

	var lastErr error
	for sender, key := range keys {
		sessionID := sessionIDs[sender]
		plaintext, err := decryptBytesAAD(key, ciphertext, []byte("yori/message/v1|"+sessionID))
		if err != nil {
			lastErr = err
			continue
		}
		unpadded, err := unpad256(plaintext)
		if err != nil {
			lastErr = err
			continue
		}
		var envelope chatEnvelope
		if err := json.Unmarshal(unpadded, &envelope); err != nil {
			lastErr = err
			continue
		}
		if envelope.Version != protocolVersion || envelope.Sender == "" || envelope.SessionID != sessionID || envelope.MessageID == "" || envelope.Counter == 0 {
			lastErr = fmt.Errorf("message sender missing")
			continue
		}
		state.mu.Lock()
		seen := state.received[sessionID]
		if seen == nil {
			seen = make(map[uint64]bool)
			state.received[sessionID] = seen
		}
		if seen[envelope.Counter] {
			state.mu.Unlock()
			return chatEnvelope{}, fmt.Errorf("replayed message")
		}
		seen[envelope.Counter] = true
		state.mu.Unlock()
		return envelope, nil
	}
	if lastErr == nil {
		lastErr = fmt.Errorf("no session keys available")
	}
	return chatEnvelope{}, lastErr
}

func decryptSessionReply(state *clientState, ciphertext string) (string, string, string, string, error) {
	state.mu.RLock()
	keys := make(map[string][]byte, len(state.shared))
	for sender, key := range state.shared {
		keys[sender] = append([]byte(nil), key...)
	}
	state.mu.RUnlock()

	var lastErr error
	for _, key := range keys {
		plaintext, err := decryptBytesAAD(key, ciphertext, []byte("yori/session/v1/reply"))
		if err != nil {
			lastErr = err
			continue
		}
		var reply sessionEnvelope
		if err := json.Unmarshal(plaintext, &reply); err != nil || reply.Sender == "" {
			lastErr = fmt.Errorf("invalid session reply")
			continue
		}
		if reply.SessionID == "" {
			lastErr = fmt.Errorf("session id missing")
			continue
		}
		return reply.Sender, reply.SessionID, reply.PublicKey, reply.Signature, nil
	}
	if lastErr == nil {
		lastErr = fmt.Errorf("no pending session keys")
	}
	return "", "", "", "", lastErr
}

func clearSession(state *clientState, recipient string) {
	state.mu.Lock()
	defer state.mu.Unlock()
	if k, ok := state.keys[recipient]; ok {
		for i := range k {
			k[i] = 0
		}
		delete(state.keys, recipient)
	}
	if k, ok := state.outgoing[recipient]; ok {
		for i := range k {
			k[i] = 0
		}
		delete(state.outgoing, recipient)
	}
	if k, ok := state.shared[recipient]; ok {
		for i := range k {
			k[i] = 0
		}
		delete(state.shared, recipient)
	}
	if sid, ok := state.sessionIDs[recipient]; ok {
		delete(state.received, sid)
		delete(state.sessionIDs, recipient)
	}
	delete(state.sendCounts, recipient)
}

func clearAllSessions(state *clientState) {
	state.mu.Lock()
	defer state.mu.Unlock()
	for u, k := range state.keys {
		for i := range k {
			k[i] = 0
		}
		delete(state.keys, u)
	}
	for u, k := range state.outgoing {
		for i := range k {
			k[i] = 0
		}
		delete(state.outgoing, u)
	}
	for u, k := range state.shared {
		for i := range k {
			k[i] = 0
		}
		delete(state.shared, u)
	}
	state.sessionIDs = make(map[string]string)
	state.sendCounts = make(map[string]uint64)
	state.received = make(map[string]map[uint64]bool)
}

File: server/web/src/session.js
// Session management, key agreement, anti-replay, and messaging
import { generateX25519KeyPair, deriveSharedSecret } from './crypto/keyagreement.js';
import { deriveDirectionalKeys, derivePreSessionKey } from './crypto/kdf.js';
import { encryptBytesAAD, decryptBytesAAD } from './crypto/encryption.js';
import { signSessionMetadata, verifySessionMetadata, getFingerprintHex } from './crypto/identity.js';
import { buildOnionPacket } from './crypto/onion.js';
import { createPacket, randomID } from './protocol.js';

export class SessionManager {
  constructor(selfUsername, identityKeyInfo, preSessionKeyInfo, sendPacket, circuitManager, logger = console.log) {
    this.selfUsername = selfUsername;
    this.identityKeyInfo = identityKeyInfo;       // { keyPair, pubB64, fingerprint }
    this.preSessionKeyInfo = preSessionKeyInfo;   // { keyPair, pubB64 }
    this.sendPacket = sendPacket;
    this.circuitManager = circuitManager;
    this.logger = logger;

    // Per-peer active session state
    // peerUsername -> { sessionID, sendKey, recvKey, sendCounter, highestRecvCounter, ephemeralKeyPair, state }
    this.sessions = new Map();

    // Fingerprint store for TOFU (peerUsername -> { fingerprint, verified: bool })
    this.fingerprints = new Map();
    this.keyChanged = new Set();
    this.peerPreSessionKeys = new Map();
    this.loadCachedFingerprints();
    this.loadPeerPreSessionKeys();
  }

  loadCachedFingerprints() {
    try {
      const raw = localStorage.getItem('yori_fingerprints');
      if (raw) {
        const parsed = JSON.parse(raw);
        for (const [k, v] of Object.entries(parsed)) {
          this.fingerprints.set(k, v);
        }
      }
    } catch (e) {}
  }

  saveCachedFingerprints() {
    try {
      const obj = {};
      for (const [k, v] of this.fingerprints.entries()) {
        obj[k] = v;
      }
      localStorage.setItem('yori_fingerprints', JSON.stringify(obj));
    } catch (e) {}
  }

  loadPeerPreSessionKeys() {
    try {
      const parsed = JSON.parse(localStorage.getItem('yori_peer_session_keys') || '{}');
      for (const [username, key] of Object.entries(parsed)) this.peerPreSessionKeys.set(username, key);
    } catch (_) {}
  }

  savePeerPreSessionKeys() {
    try {
      localStorage.setItem('yori_peer_session_keys', JSON.stringify(Object.fromEntries(this.peerPreSessionKeys)));
    } catch (_) {}
  }

  getPeerPreSessionKey(peerUsername) {
    return this.peerPreSessionKeys.get(peerUsername) || '';
  }

  setPeerPreSessionKey(peerUsername, publicKey) {
    const key = publicKey.trim();
    try {
      const base64 = key.replace(/-/g, '+').replace(/_/g, '/');
      const decoded = Uint8Array.from(atob(base64 + '='.repeat((4 - base64.length % 4) % 4)), c => c.charCodeAt(0));
      if (decoded.length !== 32) return false;
    } catch (_) {
      return false;
    }
    this.peerPreSessionKeys.set(peerUsername, key);
    this.savePeerPreSessionKeys();
    return true;
  }

  trustKey(peerUsername) {
    this.fingerprints.delete(peerUsername);
    this.keyChanged.delete(peerUsername);
    this.saveCachedFingerprints();
    this.logger(`[identity] cleared cached key for ${peerUsername} — next session offer will be accepted`);
  }

  verifyFingerprint(peerUsername) {
    const entry = this.fingerprints.get(peerUsername);
    if (entry) {
      entry.verified = true;
      this.saveCachedFingerprints();
      this.logger(`[identity] verified fingerprint for ${peerUsername}`);
    }
  }

  hasSession(peerUsername) {
    const s = this.sessions.get(peerUsername);
    return s && s.state === 'ESTABLISHED';
  }

  // Alice initiates session offer to Bob
  async startSession(targetPeer) {
    const peerUsername = targetPeer.username;
    const targetPub = this.getPeerPreSessionKey(peerUsername);
    if (!targetPub) {
      this.logger(`[error] cannot start session with ${peerUsername}: missing pre-session public key`);
      return;
    }

    const ephemeral = await generateX25519KeyPair();
    const sessionID = randomID();
    const signature = await signSessionMetadata(
      this.identityKeyInfo.keyPair.privateKey,
      this.selfUsername,
      sessionID,
      ephemeral.pubB64
    );

    // Save pending session state for Alice
    this.sessions.set(peerUsername, {
      sessionID,
      ephemeralKeyPair: ephemeral.keyPair,
      ephemeralPubB64: ephemeral.pubB64,
      sendCounter: 0,
      highestRecvCounter: 0,
      state: 'PENDING_OFFER'
    });

    const offerPayload = {
      type: 'session_offer',
      sender: this.selfUsername,
      session_id: sessionID,
      ephemeral_pub: ephemeral.pubB64,
      identity_pub: this.identityKeyInfo.pubB64,
      signature
    };

    // Encrypt offer using recipient's presession_pub
    const ephOuter = await generateX25519KeyPair();
    const shared = await deriveSharedSecret(ephOuter.keyPair.privateKey, targetPub);
    const aesKey = await derivePreSessionKey(shared);
    const ciphertextB64 = await encryptBytesAAD(
      aesKey,
      new TextEncoder().encode(JSON.stringify(offerPayload)),
      'yori-presession-v1'
    );

    // Send deliver packet to server
    const deliverPacket = createPacket('deliver', {
      to: peerUsername,
      payload: ciphertextB64,
      publicKey: ephOuter.pubB64
    });

    this.sendPacket(deliverPacket);
    this.logger(`[session] sent session offer to ${peerUsername}`);
  }

  // Bob receives and processes a session offer
  async handleSessionOffer(offer) {
    const sender = offer.sender;
    const fingerprint = await getFingerprintHex(offer.identity_pub);

    // TOFU check
    const cached = this.fingerprints.get(sender);
    if (cached && cached.fingerprint !== fingerprint) {
      this.keyChanged.add(sender);
      this.logger(`[warn] session offer from ${sender} blocked: identity key changed`);
      // Send rejection
      const rejectPayload = {
        type: 'session_reject',
        sender: this.selfUsername,
        reason: 'identity_key_changed'
      };
      await this.sendPreSessionDeliver(sender, offer.ephemeral_pub, rejectPayload);
      return;
    }

    // Verify signature
    const valid = await verifySessionMetadata(
      offer.identity_pub,
      sender,
      offer.session_id,
      offer.ephemeral_pub,
      offer.signature
    );
    if (!valid) {
      this.logger(`[warn] session offer from ${sender} blocked: invalid signature`);
      return;
    }

    // Record fingerprint if new
    if (!cached) {
      this.fingerprints.set(sender, { fingerprint, verified: false });
      this.saveCachedFingerprints();
    }

    // Bob generates own ephemeral X25519 key pair
    const ephemeral = await generateX25519KeyPair();
    const sharedSecret = await deriveSharedSecret(ephemeral.keyPair.privateKey, offer.ephemeral_pub);

    // Bob is responder: send = yori-recv-v1, recv = yori-send-v1
    const keys = await deriveDirectionalKeys(sharedSecret);
    const bobSendKey = keys.recvKey;
    const bobRecvKey = keys.sendKey;

    this.sessions.set(sender, {
      sessionID: offer.session_id,
      sendKey: bobSendKey,
      recvKey: bobRecvKey,
      sendCounter: 0,
      highestRecvCounter: 0,
      state: 'ESTABLISHED'
    });

    // Bob signs his reply
    const signature = await signSessionMetadata(
      this.identityKeyInfo.keyPair.privateKey,
      this.selfUsername,
      offer.session_id,
      ephemeral.pubB64
    );

    const replyPayload = {
      type: 'session_reply',
      sender: this.selfUsername,
      session_id: offer.session_id,
      ephemeral_pub: ephemeral.pubB64,
      identity_pub: this.identityKeyInfo.pubB64,
      signature
    };

    await this.sendPreSessionDeliver(sender, offer.ephemeral_pub, replyPayload);
    this.logger(`[session] session established with ${sender}`);
  }

  // Alice receives Bob's session reply
  async handleSessionReply(reply) {
    const sender = reply.sender;
    const pending = this.sessions.get(sender);
    if (!pending || pending.sessionID !== reply.session_id) {
      return;
    }

    const valid = await verifySessionMetadata(
      reply.identity_pub,
      sender,
      reply.session_id,
      reply.ephemeral_pub,
      reply.signature
    );
    if (!valid) {
      this.logger(`[warn] session reply from ${sender} signature verification failed`);
      return;
    }

    // Alice performs X25519 with Bob's ephemeral pub
    const sharedSecret = await deriveSharedSecret(pending.ephemeralKeyPair.privateKey, reply.ephemeral_pub);

    // Alice is initiator: send = yori-send-v1, recv = yori-recv-v1
    const keys = await deriveDirectionalKeys(sharedSecret);
    pending.sendKey = keys.sendKey;
    pending.recvKey = keys.recvKey;
    pending.state = 'ESTABLISHED';

    const fingerprint = await getFingerprintHex(reply.identity_pub);
    const cached = this.fingerprints.get(sender);
    if (cached && cached.fingerprint !== fingerprint) {
      this.sessions.delete(sender);
      this.keyChanged.add(sender);
      this.logger(`[warn] session reply from ${sender} blocked: identity key changed`);
      return;
    }
    if (!cached) {
      this.fingerprints.set(sender, { fingerprint, verified: false });
      this.saveCachedFingerprints();
    }

    this.logger(`[session] session established with ${sender}`);
  }

  handleSessionReject(reject) {
    const sender = reject.sender;
    if (reject.reason === 'identity_key_changed') {
      this.logger(`[warn] session rejected by ${sender}: identity key changed — ask them to run /trustkey ${this.selfUsername}`);
    } else {
      this.logger(`[warn] session rejected by ${sender}`);
    }
    this.sessions.delete(sender);
  }

  // Replies and rejections are sealed to the initiator's pending ephemeral
  // key, not to its long-lived pre-session key. This also lets the transport
  // remain type-blind: the outer packet is always just "deliver".
  async decryptPendingResponse(senderEphemeralPubB64, ciphertextB64) {
    for (const pending of this.sessions.values()) {
      if (pending.state !== 'PENDING_OFFER' || !pending.ephemeralKeyPair) continue;
      try {
        const shared = await deriveSharedSecret(pending.ephemeralKeyPair.privateKey, senderEphemeralPubB64);
        const key = await derivePreSessionKey(shared);
        const plaintext = await decryptBytesAAD(key, ciphertextB64, 'yori-presession-v1');
        const payload = JSON.parse(new TextDecoder().decode(plaintext));
        if (payload.type === 'session_reply' || payload.type === 'session_reject') return payload;
      } catch (_) {
        // The packet may instead be an offer/cover sealed to our pre-session key.
      }
    }
    return null;
  }

  // Send an encrypted chat message to Bob
  async sendMessage(peerUsername, bodyText) {
    const session = this.sessions.get(peerUsername);
    if (!session || session.state !== 'ESTABLISHED') {
      throw new Error(`No established session with ${peerUsername}`);
    }

    session.sendCounter += 1;
    const counter = session.sendCounter;

    const messageEnvelope = {
      type: 'message',
      session_id: session.sessionID,
      counter,
      body: bodyText,
      is_cover: false
    };

    const aad = `yori/message/v1|${session.sessionID}`;
    const ciphertextB64 = await encryptBytesAAD(
      session.sendKey,
      new TextEncoder().encode(JSON.stringify(messageEnvelope)),
      aad
    );

    // Prefer multi-hop onion routing whenever two relay peers are available.
    const circuit = this.circuitManager.getCircuit(peerUsername);
    if (!circuit || !circuit.route || circuit.route.length < 2) {
      // The server receives only this encrypted envelope; plaintext and
      // session keys remain in the participating browsers.
      this.sendPacket(createPacket('deliver', { to: peerUsername, payload: ciphertextB64 }));
      return { counter, circuit: null };
    }

    const { circuitID, targetRelay, onionPayload } = await buildOnionPacket(
      circuit.route,
      peerUsername,
      ciphertextB64,
      circuit.circuitID
    );

    const deliverPkt = createPacket('deliver', {
      to: targetRelay,
      payload: onionPayload,
      circuit_id: circuitID
    });

    this.sendPacket(deliverPkt);
    return {
      counter,
      circuit: {
        circuitID,
        path: [this.selfUsername, circuit.route[0].username, circuit.route[1].username, peerUsername]
      }
    };
  }

  // Decrypt incoming message
  async decryptMessage(sender, ciphertextB64) {
    const session = this.sessions.get(sender);
    if (!session || !session.recvKey) {
      throw new Error(`No session key to decrypt message from ${sender}`);
    }

    const aad = `yori/message/v1|${session.sessionID}`;
    let decryptedBytes;
    try {
      decryptedBytes = await decryptBytesAAD(session.recvKey, ciphertextB64, aad);
    } catch (err) {
      this.logger(`[tamper] AES-GCM authentication failure from ${sender}: message was modified or corrupted!`);
      throw err;
    }

    const envelope = JSON.parse(new TextDecoder().decode(decryptedBytes));
    if (envelope.is_cover) {
      return null; // Cover traffic silently discarded
    }

    // Replay protection: counter must be strictly greater than highest seen
    if (envelope.counter <= session.highestRecvCounter) {
      this.logger(`[warn] replay rejected from ${sender}: counter ${envelope.counter} already seen`);
      return null;
    }

    session.highestRecvCounter = envelope.counter;
    return envelope;
  }

  async sendPreSessionDeliver(toUsername, recipientPubB64, payloadObj) {
    const eph = await generateX25519KeyPair();
    const shared = await deriveSharedSecret(eph.keyPair.privateKey, recipientPubB64);
    const key = await derivePreSessionKey(shared);
    const ciphertextB64 = await encryptBytesAAD(
      key,
      new TextEncoder().encode(JSON.stringify(payloadObj)),
      'yori-presession-v1'
    );
    const packet = createPacket('deliver', {
      to: toUsername,
      payload: ciphertextB64,
      publicKey: eph.pubB64
    });
    this.sendPacket(packet);
  }
}
